src/app/missing_subchain/block.ml

```diff
@@ -0,0 +1,253 @@
+let user_cmds_tbl : Extensional.User_command.t list State_hash.Table.t =
+  State_hash.Table.create ()
+
+(* the internal commands in a block with a given state hash *)
+let internal_cmds_tbl : Extensional.Internal_command.t list State_hash.Table.t
+    =
+  State_hash.Table.create ()
+
+module Fee_transfer_via_coinbase = struct
+  (* table of fee transfer via coinbase at given global slot, seq no, secondary seq no *)
+  module T = struct
+    type t = int64 * int * int
+    [@@deriving bin_io_unversioned, hash, compare, sexp]
+  end
+
+  include T
+  include Hashable.Make_binable (T)
+end
+
+let fee_transfer_tbl :
+    Extensional.Internal_command.t Fee_transfer_via_coinbase.Table.t =
+  Fee_transfer_via_coinbase.Table.create ()
+
+let blockchain_state_of_extensional_block (block : Extensional.Block.t) =
+  {timestamp= block.timestamp}
+
+let mk_global_slot slot_number = {slot_number}
+
+let mk_ledger hash = {hash}
+
+let mk_epoch_data seed hash = {ledger= mk_ledger hash; seed}
+
+let consensus_state_of_extensional_block (block : Extensional.Block.t) :
+    consensus_state =
+  { curr_global_slot= mk_global_slot block.global_slot
+  ; global_slot_since_genesis= block.global_slot_since_genesis
+  ; staking_epoch_data=
+      mk_epoch_data block.staking_epoch_seed block.staking_epoch_ledger_hash
+  ; next_epoch_data=
+      mk_epoch_data block.next_epoch_seed block.next_epoch_ledger_hash
+  ; block_stake_winner= block.block_winner
+  ; block_creator= block.creator }
+
+let protocol_state_body_of_extensional_block (block : Extensional.Block.t) =
+  let blockchain_state = blockchain_state_of_extensional_block block in
+  let consensus_state = consensus_state_of_extensional_block block in
+  {blockchain_state; consensus_state}
+
+let protocol_state_of_extensional_block block =
+  {body= protocol_state_body_of_extensional_block block}
+
+let common_of_user_cmd (user_cmd : Extensional.User_command.t) =
+  { fee= user_cmd.fee
+  ; fee_token= user_cmd.fee_token
+  ; fee_payer_pk= user_cmd.fee_payer
+  ; nonce= user_cmd.nonce
+  ; valid_until= user_cmd.valid_until
+  ; memo= user_cmd.memo }
+
+let body_of_user_cmd (user_cmd : Extensional.User_command.t) =
+  let payload =
+    { source_pk= user_cmd.source
+    ; receiver_pk= user_cmd.receiver
+    ; token_id= user_cmd.token
+    ; amount= user_cmd.amount }
+  in
+  match user_cmd.typ with
+  | "payment" ->
+      Payment payload
+  | "delegation" ->
+      Stake_delegation payload
+  | "create_token" ->
+      Create_new_token payload
+  | "create_account" ->
+      Create_token_account payload
+  | "mint_tokens" ->
+      Mint_tokens payload
+  | cmd ->
+      failwithf "Unknown user command \"%s\"" cmd ()
+
+let payload_of_user_cmd user_cmd =
+  {common= common_of_user_cmd user_cmd; body= body_of_user_cmd user_cmd}
+
+let status_of_user_cmd (user_cmd : Extensional.User_command.t) =
+  let open Transaction_status in
+  (* balance data here is always empty, it's a dummy value
```

Comment:
Shouldn't this be a TODO? I would have expected that this will use the balance data added in #7114.